LambdaEngine/Include/Application/API/Events/EventQueue.h

```diff
@@ -4,259 +4,91 @@
 
 #include "Containers/TUniquePtr.h"
 
+#include "Memory/API/Malloc.h"
+#include "Memory/API/StackAllocator.h"
+
 namespace LambdaEngine
 {
 	/*
 	* EventContainer
 	*/
-	struct EventContainer
-	{
-	public:
-		using Iterator = Event*;
-		using ConstIterator = const Event*;
-
-		virtual ~EventContainer() = default;
-
-		inline EventContainer(EventType eventType) noexcept
-			: ContainerEventType(eventType)
-		{
-		}
 
-		virtual void Push(const Event& event) = 0;
-		virtual void Clear() = 0;
-		
-		virtual uint32 Size() const = 0;
-
-		virtual EventContainer* Copy(void* pMemory) = 0;
-		virtual EventContainer* Move(void* pMemory) = 0;
-		
-		virtual Event& GetAt(uint32 index) = 0;
-		virtual const Event& GetAt(uint32 index) const = 0;
-
-	public:
-		EventType ContainerEventType;
-	};
-
-	/*
-	* TEventContainer
-	*/
-	template<typename TEvent>
-	struct TEventContainer : public EventContainer
+	class EventContainer
 	{
 	public:
-		inline TEventContainer(uint32 size = 1) noexcept
-			: EventContainer(TEvent::GetStaticType())
-			, Events()
-		{
-			Events.Reserve(size);
-		}
-
-		inline TEventContainer(const TEventContainer& other) noexcept
-			: EventContainer(TEvent::GetStaticType())
-			, Events(other.Events)
+		inline EventContainer()
+			: m_Allocator()
+			, m_Events()
 		{
 		}
 
-		inline TEventContainer(TEventContainer&& other) noexcept
-			: EventContainer(TEvent::GetStaticType())
-			, Events(std::move(other.Events))
-		{
-		}
-
-		virtual void Push(const Event& event) override final
-		{
-			VALIDATE(ContainerEventType == event.GetType());
-			Events.EmplaceBack(EventCast<TEvent>(event));
-		}
+		~EventContainer() = default;
 
-		virtual void Clear() override final
-		{
-			Events.Clear();
-		}
-
-		virtual uint32 Size() const override final
-		{
-			return static_cast<uint32>(Events.GetSize());
-		}
-
-		virtual EventContainer* Copy(void* pMemory)
+		template<typename TEvent>
+		FORCEINLINE void Push(const TEvent& event)
 		{
-			new(pMemory) TEventContainer(*this);
-			return reinterpret_cast<EventContainer*>(pMemory);
+			void* pMemory = m_Allocator.Push(sizeof(TEvent));
+			new(pMemory) TEvent(event);
+			m_Events.EmplaceBack(reinterpret_cast<TEvent*>(pMemory));
 		}
 
-		virtual EventContainer* Move(void* pMemory)
+		FORCEINLINE void Pop()
 		{
-			new(pMemory) TEventContainer(std::move(*this));
-			return reinterpret_cast<EventContainer*>(pMemory);
-		}
+			// Call destructor
+			m_Events.GetBack()->~Event();
```

Comment:
Why call it explicitly? Doesn't `PopBack` call the destructor?

Reply:
Since the m_Events only store pointers to events (i.e Event*). This is because the allocator owns the memory and the TArray only stores the pointers to these in form of the base class. The PopBack will not call the destructor since pointers are trivial types. All this is done because we need to have all events in one queue instead of multiple (one per type) as we had before. There probably are another solution to this, but this seems to be the most efficient.